src/query/api/v1/handler/namespace/mark_ready.go

```diff
@@ -0,0 +1,227 @@
+// Copyright (c) 2020  Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package namespace
+
+import (
+	"fmt"
+	"net/http"
+	"path"
+	"time"
+
+	clusterclient "github.com/m3db/m3/src/cluster/client"
+	"github.com/m3db/m3/src/dbnode/client"
+	nsproto "github.com/m3db/m3/src/dbnode/generated/proto/namespace"
+	"github.com/m3db/m3/src/dbnode/namespace"
+	"github.com/m3db/m3/src/dbnode/storage/index"
+	"github.com/m3db/m3/src/m3ninx/idx"
+	"github.com/m3db/m3/src/query/api/v1/handler"
+	"github.com/m3db/m3/src/query/api/v1/handler/prometheus/handleroptions"
+	"github.com/m3db/m3/src/query/generated/proto/admin"
+	"github.com/m3db/m3/src/query/storage/m3"
+	"github.com/m3db/m3/src/query/util/logging"
+	"github.com/m3db/m3/src/x/ident"
+	"github.com/m3db/m3/src/x/instrument"
+	xhttp "github.com/m3db/m3/src/x/net/http"
+	"github.com/m3db/m3/src/x/retry"
+
+	"github.com/gogo/protobuf/jsonpb"
+	"go.uber.org/zap"
+)
+
+var (
+	// M3DBMarkReadyURL is the url for the M3DB namespace add handler.
+	M3DBMarkReadyURL = path.Join(handler.RoutePrefixV1, M3DBServiceNamespacePathName, "mark_ready")
+
+	// MarkReadyHTTPMethod is the HTTP method used with this resource.
+	MarkReadyHTTPMethod = http.MethodPost
+
+	fetchRetryConfig = retry.Configuration{
+		InitialBackoff: time.Millisecond * 500,
+		MaxBackoff:     time.Second * 2,
+		BackoffFactor:  2,
+		MaxRetries:     2,
+	}
+)
+
+// MarkReadyHandler is the handler for marking namespaces ready.
+type MarkReadyHandler Handler
```

Comment:
Ugh this `Handler` was a bit confusing; know it's not really part of this PR, but any chance that can be renamed to `namespaceHandler`? (think we can get away without exporting it)

Reply:
Hm actually reading it a bit more we have annoying stuff like `nsRegistry, err = h.namespaceAddHandler.Add(namespaceRequest, opts)` which requires a base `AddHandler` type, probably fine to leave this since it's a little annoying to add interfaces to all of these?